src/main/java/org/mule/extension/ftp/internal/operation/BaseFileSystemOperations.java

```diff
@@ -0,0 +1,343 @@
+   * after their parent directory.
+   * <p>
+   *
+   * @param config        the config that is parameterizing this operation
+   * @param directoryPath the path to the directory to be listed
+   * @param recursive     whether to include the contents of sub-directories. Defaults to false.
+   * @param matchWith     a matcher used to filter the output list
+   * @param subsetList    parameter group that lets you obtain a subset of the results
+   * @return a {@link PagingProvider} of {@link Result} objects each one containing each file's content in the payload and metadata in the
+   * attributes
+   * @throws IllegalArgumentException if {@code directoryPath} points to a file which doesn't exist or is not a directory
+   */
+  protected PagingProvider<FileSystem, Result<String, FtpFileAttributes>> doPagedList(FileConnectorConfig config,
+                                                                                      String directoryPath,
+                                                                                      boolean recursive,
+                                                                                      FileMatcher matchWith,
+                                                                                      StreamingHelper streamingHelper,
+                                                                                      SubsetList subsetList) {
+    return new PagingProvider<FileSystem, Result<String, FtpFileAttributes>>() {
+
+      private List<Result<String, FtpFileAttributes>> files;
+      private Iterator<Result<String, FtpFileAttributes>> filesIterator;
+      private final AtomicBoolean initialised = new AtomicBoolean(false);
+
+      @Override
+      public List<Result<String, FtpFileAttributes>> getPage(FileSystem connection) {
+        if (initialised.compareAndSet(false, true)) {
+          initializePagingProvider(connection);
+        }
+        List<Result<String, FtpFileAttributes>> page = new LinkedList<>();
+        for (int i = 0; i < LIST_PAGE_SIZE && filesIterator.hasNext(); i++) {
+          Result<String, FtpFileAttributes> result = filesIterator.next();
+          page.add((Result.<String, FtpFileAttributes>builder().attributes(result.getAttributes().get())
+              .output(result.getOutput())
+              .mediaType(result.getMediaType().orElse(null))
+              .attributesMediaType(result.getAttributesMediaType().orElse(null))
+              .build()));
+        }
+        return page;
+      }
+
+      private void initializePagingProvider(FileSystem connection) {
+        connection.changeToBaseDir();
+        files = connection.list(config, directoryPath, recursive, getPredicate(matchWith), subsetList);
+        filesIterator = files.iterator();
+      }
+
+      @Override
+      public java.util.Optional<Integer> getTotalResults(FileSystem connection) {
+        return java.util.Optional.of(files.size());
+      }
```

Comment:
Doesn't seem to be the case (so no change is required here), but please be aware of the fact that if the result of this could change over the time, it could lead to the paging mechanism to hang waiting for results that hasn't come yet. in that case you could just make it return an Optional.empty() to avoid that situation.

Reply:
added check to return Optional.empty() in case files itself is null